index.ts

```diff
@@ -432,28 +446,236 @@ export class CapabilitiesController extends BaseController<any, any> implements
    * @param {string} domainName - The grantee domain.
    * @param {Array} newPermissions - The unique, new permissions for the grantee domain.
    */
-  addPermissionsFor (domainName: string, newPermissions: { [methodName: string]: IOcapLdCapability }) {
+  addPermissionsFor (
+    domainName: string,
+    newPermissions: { [methodName: string]: IOcapLdCapability }
+  ): void {
     const domain: RpcCapDomainEntry = this.getOrCreateDomainSettings(domainName);
     const newKeys = Object.keys(newPermissions);
 
-    // remove old permissions this will be overwritten
+    // remove old permissions so that they will be overwritten
     domain.permissions = domain.permissions.filter((oldPerm: IOcapLdCapability) => {
       return !newKeys.includes(oldPerm.parentCapability);
     });
 
     for (const methodName in newPermissions) {
-      const newPerm = newPermissions[methodName];
-      
-      domain.permissions.push(new Capability({
-        method: newPerm.parentCapability,
-        invoker: domainName,
-        caveats: newPerm.caveats,
-      }));
+      domain.permissions.push(newPermissions[methodName]);
     }
 
     this.setDomain(domainName, domain);
   }
 
+  /**
+   * Validates the given caveats. Returns true if valid, false otherwise.
+   * If the caveats have names, they must be unique.
+   * 
+   * @param {IOcapLdCaveat[]} - The caveats to validate.
+   */
+  validateCaveats (caveats: IOcapLdCaveat[]): boolean {
+
+    const seenNames: { [key: string]: boolean } = {}
+    for (const c of caveats) {
+      if (
+        typeof c !== 'object' || Array.isArray(c) ||
+        !c.type || typeof c.type !== 'string' || 
```

Comment:
This if check might be more readable if a `validateCaveat(caveat: I)capLdCaveat)` method were created to handle the `typeof c !== 'object' || Array.isArray(c) || !c.type || typeof c.type !== 'string' || c.name === ''` check.

Reply:
Assuming we added a `validateCaveat` method, we might further improve readability with the below rewrite

```
validateCaveats (caveats: IOcapLdCaveat[]): boolean {
    const allCaveatsAreValid = caveats.every(validateCaveat)
    const caveatNames = caveats.map(caveat => caveat.name)
    const allCaveatsNamesAreUnique = caveatNames.length === new Set(caveatNames).size
    return allCaveatsAreValid && allCaveatsNamesAreUnique
}
```

Reply:
This is an awesome suggestion. Because of our performance concerns I kept the single `for... of` loop construction in `validateCaveats`,* but did add the `validateCaveat` class method.

*By my count, in your more readable construction, `caveats` are iterated over thrice and two new objects are instantiated, instead of one iteration and one object in the `for... of` construction.